mllib/src/main/scala/org/apache/spark/mllib/tree/GradientBoostedTrees.scala

```diff
@@ -177,102 +177,108 @@ object GradientBoostedTrees extends Logging {
+          // Stop training early if
+          // 1. Reduction in error is less than the validationTol or
+          // 2. If the error increases, that is if the model is overfit.
+          // We want the model returned corresponding to the best validation error.
 
-        validatePredError = GradientBoostedTreesModel.updatePredictionError(
-          validationInput, validatePredError, baseLearnerWeights(m), baseLearners(m), loss)
-        val currentValidateError = validatePredError.values.mean()
-        if (bestValidateError - currentValidateError < validationTol) {
-          return new GradientBoostedTreesModel(
-            boostingStrategy.treeStrategy.algo,
-            baseLearners.slice(0, bestM),
-            baseLearnerWeights.slice(0, bestM))
-        } else if (currentValidateError < bestValidateError) {
-            bestValidateError = currentValidateError
-            bestM = m + 1
+          validatePredError = GradientBoostedTreesModel.updatePredictionError(
+            validationInput, validatePredError, baseLearnerWeights(m), baseLearners(m), loss)
+          val currentValidateError = validatePredError.values.mean()
+          if (bestValidateError - currentValidateError < validationTol) {
+            return new GradientBoostedTreesModel(
+              boostingStrategy.treeStrategy.algo,
+              baseLearners.slice(0, bestM),
+              baseLearnerWeights.slice(0, bestM))
+          } else if (currentValidateError < bestValidateError) {
+              bestValidateError = currentValidateError
+              bestM = m + 1
+          }
         }
+        // Update data with pseudo-residuals
+        data = predError.zip(input).map { case ((pred, _), point) =>
+          LabeledPoint(-loss.gradient(pred, point.label), point.features)
+        }
+        m += 1
       }
-      // Update data with pseudo-residuals
-      data = predError.zip(input).map { case ((pred, _), point) =>
-        LabeledPoint(-loss.gradient(pred, point.label), point.features)
-      }
-      m += 1
-    }
 
-    timer.stop("total")
+      timer.stop("total")
+
+      logInfo("Internal timing for DecisionTree:")
+      logInfo(s"$timer")
+      if (validate) {
+        new GradientBoostedTreesModel(
+          boostingStrategy.treeStrategy.algo,
+          baseLearners.slice(0, bestM),
+          baseLearnerWeights.slice(0, bestM))
+      } else {
+        new GradientBoostedTreesModel(
+          boostingStrategy.treeStrategy.algo, baseLearners, baseLearnerWeights)
+      }
 
-    logInfo("Internal timing for DecisionTree:")
-    logInfo(s"$timer")
-    if (validate) {
-      new GradientBoostedTreesModel(
-        boostingStrategy.treeStrategy.algo,
-        baseLearners.slice(0, bestM),
-        baseLearnerWeights.slice(0, bestM))
-    } else {
-      new GradientBoostedTreesModel(
-        boostingStrategy.treeStrategy.algo, baseLearners, baseLearnerWeights)
+    } finally {
+       if (persistedInput) input.unpersist()
```

Comment:
I agree it's cleaner this way, if a little more complex.  Concerning consistency, let me ping @mengxr since, if we do this here, we might start talking about doing it everywhere.